src/node_i18n.cc

```diff
@@ -62,6 +72,296 @@ bool flag_icu_data_dir = false;
+
+  len = ucnv_fromUChars(to.conv, *buf, length, source, length, &status);
+
+  if (U_SUCCESS(status)) {
+    args.GetReturnValue().Set(
+      AsBuffer(env->isolate(), &buf, len).ToLocalChecked());
+  } else {
+    ThrowICUError(env, status, "Unable to transcode buffer");
+  }
+}
+
+// Converts from UTF-8 to UCS2
+// args[0] must be a buffer instance
+void Ucs2FromUtf8(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[0]);
+  SPREAD_ARG(args[0], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  MaybeStackBuffer<UChar> buf;
+  int32_t len;
+
+  u_strFromUTF8(*buf, 1024, &len, ts_obj_data, ts_obj_length, &status);
+
+  if (U_SUCCESS(status)) {
+    args.GetReturnValue().Set(
+        AsBuffer(env->isolate(), &buf, len * sizeof(UChar)).ToLocalChecked());
+  } else if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    buf.AllocateSufficientStorage(len);
+    u_strFromUTF8(*buf, len, &len, ts_obj_data, ts_obj_length, &status);
+    if (U_FAILURE(status)) {
+      ThrowICUError(env, status, "Unable to transcode buffer");
+      return;
+    }
+    args.GetReturnValue().Set(
+      AsBuffer(env->isolate(), &buf, len << 1).ToLocalChecked());
+  } else {
+    ThrowICUError(env, status, "Unable to transcode buffer");
+  }
+}
+
+// Converts UCS2 into UTF-8
+// args[0] must be a buffer instance
+void Utf8FromUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[0]);
+  SPREAD_ARG(args[0], ts_obj);
+
+  const size_t length = ts_obj_length >> 1;
+
+  UChar* source = nullptr;
+  MaybeStackBuffer<UChar> swapspace;
+  if (IsLittleEndian()) {
+    source = reinterpret_cast<UChar*>(ts_obj_data);
+  } else {
+    swapspace.AllocateSufficientStorage(length);
+    for (size_t n = 0, i = 0; i < length; n += 2, i += 1) {
+      const uint8_t hi = static_cast<uint8_t>(ts_obj_data[n + 0]);
+      const uint8_t lo = static_cast<uint8_t>(ts_obj_data[n + 1]);
+      swapspace[i] = (hi << 8) | lo;
+    }
+    source = *swapspace;
+  }
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  MaybeStackBuffer<char> buf;
+  int32_t len;
+
+  u_strToUTF8(*buf, 1024, &len, source, length, &status);
+  if (U_SUCCESS(status)) {
+    // Used the static buf, just copy it into the new Buffer instance.
+    args.GetReturnValue().Set(
+      AsBuffer(env->isolate(), &buf, len).ToLocalChecked());
+  } else if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    buf.AllocateSufficientStorage(len);
+    u_strToUTF8(*buf, len, &len, source, length, &status);
+    if (U_FAILURE(status)) {
+      ThrowICUError(env, status, "Unable to transcode buffer");
+      return;
+    }
+    args.GetReturnValue().Set(
+      Buffer::New(env->isolate(), *buf, len).ToLocalChecked());
+    buf.Release();
+  } else {
+    ThrowICUError(env, status, "Unable to transcode buffer");
+  }
+}
```

Comment:
It seems like there is ample opportunity to share code between Ucs2FromUtf8 and Utf8FromUcs2, they are 80% identical.